### PR TITLE
Add .duration data type

### PR DESCRIPTION
**Pull request by syoung-smallwisdom**
This includes adding the `.duration` data type to the form data types *and* removing the `.timeInterval` from the base answer types. I removed the time interval data type b/c it can be described using a `.decimal` type and thus it was redundant. Adding the time interval type also highlighted the issue of the picker fields not properly displaying the separator string (if there is one) so I fixed that as well.

### CODE REVIEW

File: ResearchSuite/ResearchSuite/RSDDurationPickerDataSourceObject.swift [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+}
+
+extension RSDDurationRange {
+    
+    /// Convenience method for getting the base unit associated with this range.
+    public var baseUnit: UnitDuration {
+        return self.minimumDuration.unit
+    }
+    
+    /// Convenience method for getting the date components formatter for this range
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDDurationPickerDataSourceObject.swift]: needs a period at the end

File: ResearchSuite/ResearchSuite/RSDFormDataType.swift [hunk 2]
```diff
@@ -69,33 +69,31 @@ public enum RSDFormDataType {
...
-        
         /// The decimal question type asks the participant to enter a decimal number. A decimal data type can
         /// map to a `RSDNumberRange` to box the allowed values.
         case decimal
         
-        /// The integer question type asks the participant to enter an integer number. An integer data type can
-        /// map to a `RSDNumberRange` to box the allowed values, but will store the value as an `Int`.
-        case integer
+        /// In a duration question, the participant can enter a time span such as "8 hours, 5 minutes"
+        /// or "3 minutes, 15 seconds"
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDFormDataType.swift]: Needs a period at the end

File: ResearchSuite/ResearchSuite/RSDInputFieldObject.swift [hunk 3]
```diff
@@ -153,11 +153,12 @@ open class RSDInputFieldObject : RSDSurveyInputField, Codable {
...
-    /// |--------------------------|:----------------------------------------------------------:|
-    /// | .integer, .decimal       | `RSDNumberRangeObject`                                     |
-    /// | .date                    | `RSDDateRangeObject`                                       |
-    /// | .year                    | `RSDDateRangeObject` or `RSDNumberRangeObject`             |
+    /// | RSDFormDataType.BaseType      | Type of range to decode                                    |
+    /// |-------------------------------|:----------------------------------------------------------:|
+    /// | .integer, .decimal, .fraction | `RSDNumberRangeObject`                                     |
+    /// | .date                         | `RSDDateRangeObject`                                       |
+    /// | .year                         | `RSDDateRangeObject` or `RSDNumberRangeObject`             |
+    /// | .timeInterval                 | `RSDDurationRangeObject`                               |
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDInputFieldObject.swift]: alignment of the final '|' with the lines above

File: ResearchSuiteUI/ResearchSuiteUI/iOS/Views/RSDPickerViewProtocol.swift [hunk 4]
```diff
@@ -129,65 +129,227 @@ open class RSDChoicePickerView : UIPickerView, RSDPickerViewProtocol, UIPickerVi
...
+    /// method can cause returning `0` or an infinite loop of wacky madness.
+    open override var numberOfComponents: Int {
+        if hasSeparator {
+            return pickerSource.numberOfComponents * 2 - 1
+        } else {
+            return pickerSource.numberOfComponents
+        }
+    }
+    
+    /// Override super to return the number of rows for each comonent returned by the picker
```
Comment (Erin-Mounts) [inline ResearchSuiteUI/ResearchSuiteUI/iOS/Views/RSDPickerViewProtocol.swift]: comonent --> component

File: ResearchSuiteUI/ResearchSuiteUI/iOS/Views/RSDPickerViewProtocol.swift [hunk 5]
```diff
@@ -129,65 +129,227 @@ open class RSDChoicePickerView : UIPickerView, RSDPickerViewProtocol, UIPickerVi
...
+    
+    /// Transform the component into the components for the picker source, or return `nil`
+    /// if this is a separator row.
+    func transformedComponent(_ component: Int) -> Int? {
+        guard let idx = pickerComponents().index(of: component) else { return nil }
+        return idx
+    }
+    
+    /// Returns the array of components that map to the components of the picker.
+    /// This is needed because the separator is displayed using it's own row.
```
Comment (Erin-Mounts) [inline ResearchSuiteUI/ResearchSuiteUI/iOS/Views/RSDPickerViewProtocol.swift]: it's --> its

File: ResearchSuiteUI/ResearchSuiteUI/iOS/Views/RSDPickerViewProtocol.swift [hunk 6]
```diff
@@ -129,65 +129,227 @@ open class RSDChoicePickerView : UIPickerView, RSDPickerViewProtocol, UIPickerVi
...
+            return hasSeparator ? pickerSource.separator : nil
+        }
+        guard let choice = choice(forRow: row, forComponent: component) else {
+            return nil
+        }
         return choice.text
     }
     
+    /// Return a view to use for this picker. By default, this method returns a label that is
+    /// alligned with the components center aligned.
```
Comment (Erin-Mounts) [inline ResearchSuiteUI/ResearchSuiteUI/iOS/Views/RSDPickerViewProtocol.swift]: alligned --> aligned

File: ResearchSuite/ResearchSuite/RSDDateRangeObject.swift [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+    
+    /// The minimum allowed date. When the value of this property is `nil`, then the `allowPast`
+    /// property is checked for `nil`, otherwise `allowPast` is ignored.
+    public let minDate: Date?
+    
+    /// The maximum allowed date. When the value of this property is `nil`, then the `allowFuture`
+    /// property is checked for `nil`, otherwise `allowFuture` is ignored.
+    public let maxDate: Date?
+    
+    /// Whether or not the UI should allow future dates. If `nil` or if `minDate` is defined then this value
```
[the remark this answers is not in the document]
Comment (syoung-smallwisdom) [reply ResearchSuite/ResearchSuite/RSDDateRangeObject.swift]: whuhu?

```
extension RSDDateRange {
    
    /// The minimum allowed date. This is calculated by using either the `minDate` (if non-nil) or today's
    /// date if `allowPast` is non-nil and `false`. If both `minDate` and `allowPast` are `nil` then this
    /// property will return `nil`.
    public var minimumDate: Date? {
        return minDate ?? ((allowPast ?? true) ? nil : Date())
    }
    
    /// The maximum allowed date. This is calculated by using either the `maxDate` (if non-nil) or today's
    /// date if `allowFuture` is non-nil and `false`. If both `maxDate` and `allowFuture` are `nil` then this
    /// property will return `nil`.
    public var maximumDate: Date? {
        return maxDate ?? ((allowFuture ?? true) ? nil : Date())
    }
}
```

**Pull request merged**